util/src/main/scala/stryker4s/extensions/mutationtypes/MutationTypes.scala

```diff
@@ -28,7 +28,35 @@ trait BooleanSubstitution extends Mutation[Lit.Boolean]
 
 trait LogicalOperator extends Mutation[Term.Name]
 
-trait MethodMutator extends Mutation[Term.Name]
+// TODO: check another mutators (maybe there is a better way to match them), find a common interface
+trait MethodMutator {
```

Comment:
Let me check that one more time, I'm sure I can find a way to still extend from Mutation[T]

Reply:
I think there are at least two kind of Mutations, substitution mutations (like String and Boolean, they change a Literal for another) and method mutations (like infix, non args o one arg method calls).

The first one is like the old Mutation, they change a Tree for another Tree. They could be something like

```trait SubstitutionMutation[T <: Tree] {
  protected val tree:T
  def unapply(arg: T): Option[T] = Some(arg).filter(_.isEqual(tree))
}```

The seconds have a different behavior, they change a small part inside a Tree, so they need to provide a function instead a Tree. They could be something like

```trait MethodMutation[T <: Tree] {
  protected val methodName: String
  def apply(f: String => T): T = f(methodName)
  def unapply(term: T): Option[(T, String => T)]
}```

As you can see, there is nothing in common between both of them. So I think there are two paths,

1) let them extend from (without methods)
```trait Mutation[T <: Tree]```

2) or try to change the way substitution mutation works in order to get an interface like method mutation.

@hugo-vrijswijk , what do you think?

Reply:
Hmm, good observation of the only two distinct differences. Certainly something we could use. For now, I think we should limit this PR to method mutations only (with infix) so we can merge it. I think most errors of the original issue come from method mutations. And then later we could have a look at a more general solution

Reply:
Currently Mutation[T] is a Trait that needs a ```val tree:T```, and that interface doesn't work with MethodMutation[T]. If I extended MethodMutation[T] from Mutation[T], I would have to define a ```val tree:T``` to MethodMutation[T] implementations, a useless property for MethodMutation[T]. Also MutationMethod[T] will inherit an apply method ```def unapply(arg: T): Option[T]``` that it doesn't need it. 

It will look like this
```
trait MethodMutator[T <: Tree] extends MethodMutator[T] {
  val tree:T = ???
  protected val methodName: String
  def apply(f: String => T): T = f(methodName)
  def unapply(term: T): Option[(T, String => T)]
}
```

If we want that all mutations extend from something, a cheaper way to do that is

```
trait Mutation[T <: Tree]

trait SubstitutionMutation[T <: Tree] extends Mutation[T]  {  val tree.... } // this trait is the old Mutation[T]

trait MethodMutation[T <: Tree] extends Mutation[T]
```